File: crates/app_state/tests/graph_wire.rs
// clippy.toml's allow-unwrap-in-tests does not reach helpers beside `#[test]` fns.
#![allow(clippy::unwrap_used, clippy::expect_used)]

use app_state::{GraphWindow, graph_wire};
use git_engine::CommitRow;
use graph_engine::{GraphRow, NodeKind, Segment, Span};

/// The same bytes are decoded by `frontend/src/lib/graph-wire.test.ts`.
const GOLDEN: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../frontend/src/lib/graph-wire.golden.bin"
);

fn commit(oid: char, summary: &str, name: &str, timestamp: i64, zone: i32) -> CommitRow {
    CommitRow {
        oid: oid.to_string().repeat(40),
        parents: vec!["f".repeat(40)],
        summary: summary.to_owned(),
        author_name: name.to_owned(),
        author_email: format!("{}@example.com", name.to_lowercase()),
        timestamp,
        tz_offset_minutes: zone,
    }
}

fn segment(from: u16, to: u16, span: Span, primary: bool, arrow: bool) -> Segment {
    Segment {
        from,
        to,
        span,
        primary,
        color: from.try_into().unwrap(),
        arrow,
    }
}

fn sample() -> GraphWindow {
    GraphWindow {
        start: 40,
        total: 1000,
        complete: false,
        commits: vec![
            commit('a', "Merge branch 'feature'", "Ann", 1_700_000_000, 180),
            commit('b', "Übersicht — 修正", "Bo", 1_600_000_000, -300),
        ],
        rows: vec![
            GraphRow {
                row: 40,
                lane: 0,
                color: 0,
                kind: NodeKind::Merge,
                primary: true,
                width: 2,
                segments: vec![
                    segment(0, 0, Span::Bottom, true, false),
                    segment(0, 1, Span::Bottom, false, false),
                ],
            },
            GraphRow {
                row: 41,
                lane: 1,
                color: 3,
                kind: NodeKind::Root,
                primary: false,
                width: 2,
                segments: vec![segment(1, 1, Span::Top, false, true)],
            },
        ],
    }
}

#[test]
fn the_frontend_decodes_the_same_bytes() {
    let bytes = graph_wire::encode(&sample());
    if std::env::var_os("COGIT_BLESS").is_some() {
        std::fs::write(GOLDEN, &bytes).unwrap();
    }
    assert_eq!(
        bytes,
        std::fs::read(GOLDEN).unwrap(),
        "the format changed: bless with COGIT_BLESS=1 and update graph-wire.test.ts"
    );
}

#[test]
fn the_buffer_is_as_long_as_its_columns() {
    let window = sample();
    let bytes = graph_wire::encode(&window);

    let rows = 2;
    let segments = 3;
    let text: usize = window
        .commits
        .iter()
        .map(|c| c.summary.len() + c.author_name.len() + c.author_email.len())
        .sum();
    let columns = rows * (8 + 4 + 4 + 2 + 2 + 1 + 1 + 40) + 4 + (rows * 3 + 1) * 4;
    assert_eq!(
        bytes.len(),
        32 + columns + segments * (2 + 2 + 1 + 1) + text
    );
}

#[test]
fn an_empty_window_is_a_header() {
    let window = GraphWindow {
        start: 0,
        total: 0,
        complete: true,
        commits: Vec::new(),
        rows: Vec::new(),
    };
    // The segment and text offset columns still end with their closing offset.
    assert_eq!(graph_wire::encode(&window).len(), 32 + 4 + 4);
}

File: frontend/src/lib/graph-wire.ts
import type { CommitRow, GraphRow, Segment } from "$lib/ipc";

/** A commit as the list shows it; parents stay in Rust, the list never draws from them. */
export type GraphCommit = Omit<CommitRow, "parents">;

export interface GraphEntry {
  commit: GraphCommit;
  layout: GraphRow;
}

/** A window of the graph as it came over: columns over one buffer, decoded row by row. */
export interface GraphBlock {
  start: number;
  total: number;
  complete: boolean;
  length: number;
  oid(row: number): string;
  /** The row of `oid` in this block, or -1. */
  find(oid: string): number;
  entry(row: number): GraphEntry;
}

const VERSION = 2;
const KINDS: readonly GraphRow["kind"][] = ["normal", "merge", "root", "workingTree"];
const SPANS: readonly Segment["span"][] = ["top", "bottom", "through"];
const utf8 = new TextDecoder();

type View<T> = new (buffer: ArrayBuffer, offset: number, length: number) => T;

/** The window as it crosses `postMessage`: base64 of the columns (R-194). */
export function decodeBase64Window(text: string): GraphBlock | null {
  const native = (Uint8Array as unknown as { fromBase64?: (s: string) => Uint8Array }).fromBase64;
  if (native) return decodeWindow(native(text).buffer as ArrayBuffer);
  const raw = atob(text);
  const bytes = new Uint8Array(raw.length);
  for (let i = 0; i < raw.length; i++) bytes[i] = raw.charCodeAt(i);
  return decodeWindow(bytes.buffer);
}

/** The layout is in `crates/app_state/src/graph_wire.rs`. An empty buffer is a window
    of a graph that was replaced before it was asked for. */
export function decodeWindow(buffer: ArrayBuffer): GraphBlock | null {
  if (buffer.byteLength === 0) return null;
  const [version, start, rows, total, complete, segments, textBytes, oidBytes] = new Uint32Array(buffer, 0, 8);
  if (version !== VERSION) throw new Error(`graph window format ${version}, expected ${VERSION}`);

  let at = 32;
  function take<T>(Kind: View<T> & { BYTES_PER_ELEMENT: number }, length: number): T {
    const view = new Kind(buffer, at, length);
    at += length * Kind.BYTES_PER_ELEMENT;
    return view;
  }
  const time = take(Float64Array, rows!);
  const firstSegment = take(Uint32Array, rows! + 1);
  const firstLink = take(Uint32Array, rows! + 1);
  const links = firstLink[rows!]!;
  const textAt = take(Uint32Array, rows! * 3 + 1);
  const zone = take(Int32Array, rows!);
  const lane = take(Uint16Array, rows!);
  const width = take(Uint16Array, rows!);
  const from = take(Uint16Array, segments!);
  const to = take(Uint16Array, segments!);
  const linkSegment = take(Uint16Array, links);
  const colour = take(Uint8Array, rows!);
  const flags = take(Uint8Array, rows!);
  const segmentColour = take(Uint8Array, segments!);
  const segmentFlags = take(Uint8Array, segments!);
  const oids = take(Uint8Array, rows! * oidBytes!);
  const linkOids = take(Uint8Array, links * oidBytes!);
  const text = take(Uint8Array, textBytes!);

  const field = (row: number, which: number) =>
    utf8.decode(text.subarray(textAt[row * 3 + which], textAt[row * 3 + which + 1]));
  const oid = (row: number) => utf8.decode(oids.subarray(row * oidBytes!, (row + 1) * oidBytes!));
  const decoded: GraphEntry[] = [];
  // Hex ids are ASCII: one decode of the whole column, sliced at fixed width.
  let column: string | null = null;

  return {
    start: start!,
    total: total!,
    complete: complete === 1,
    length: rows!,
    oid,
    find(wanted) {
      if (wanted.length !== oidBytes) return -1;
      column ??= utf8.decode(oids);
      for (let at = column.indexOf(wanted); at >= 0; at = column.indexOf(wanted, at + 1)) {
        if (at % oidBytes === 0) return at / oidBytes;
      }
      return -1;
    },
    entry(row) {
      const known = decoded[row];
      if (known) return known;
      const drawn: Segment[] = [];
      for (let s = firstSegment[row]!; s < firstSegment[row + 1]!; s++) {
        const bits = segmentFlags[s]!;
        drawn.push({
          from: from[s]!,
          to: to[s]!,
          span: SPANS[bits & 3]!,
          primary: (bits & 4) !== 0,
          color: segmentColour[s]!,
          arrow: (bits & 8) !== 0,
        });
      }
      const far: GraphRow["links"] = [];
      for (let l = firstLink[row]!; l < firstLink[row + 1]!; l++) {
        far.push({ segment: linkSegment[l]!, oid: utf8.decode(linkOids.subarray(l * oidBytes!, (l + 1) * oidBytes!)) });
      }
      const entry: GraphEntry = {
        commit: {
          oid: oid(row),
          summary: field(row, 0),
          authorName: field(row, 1),
          authorEmail: field(row, 2),
          timestamp: time[row]!,
          tzOffsetMinutes: zone[row]!,
        },
        layout: {
          row: start! + row,
          lane: lane[row]!,
          color: colour[row]!,
          kind: KINDS[flags[row]! & 3]!,
          primary: (flags[row]! & 4) !== 0,
          width: width[row]!,
          segments: drawn,
          links: far,
        },
      };
      decoded[row] = entry;
      return entry;
    },
  };
}
